ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -15,52 +15,256 @@
+  public UtilizationMetricsListener(
+      final Metrics metricRegistry,
+      final Map<String, List<TaskStorageMetric>> taskMetrics) {
+    this.metricsSeen = taskMetrics;
+    this.metricRegistry = metricRegistry;
   }
 
   @Override
-  public void onCreate(
-      final ServiceContext serviceContext,
-      final MetaStore metaStore,
-      final QueryMetadata queryMetadata) {
-    kafkaStreams.add(queryMetadata.getKafkaStreams());
+  public void init(List<KafkaMetric> list) {
+
   }
 
+
   @Override
-  public void onDeregister(final QueryMetadata query) {
-    final KafkaStreams streams = query.getKafkaStreams();
-    kafkaStreams.remove(streams);
+  public void configure(Map<String, ?> map) {
+    String dir;
+    if (map.containsKey(StreamsConfig.STATE_DIR_CONFIG)) {
+      dir = map.get(StreamsConfig.STATE_DIR_CONFIG).toString();
+    } else {
+      dir =  StreamsConfig.configDef().defaultValues().get(StreamsConfig.STATE_DIR_CONFIG).toString();
+    }
+    configureShared(new File(dir), this.metricRegistry);
+  }
+
+  private static void configureShared(final File baseDir, final Metrics metricRegistry) {
+    synchronized (lock) {
+      if (registeredNodeMetrics != null) {
+        return;
+      }
+      registeredNodeMetrics = new ArrayList<>();
+
+      final MetricName nodeAvailable = metricRegistry.metricName("node-storage-available", "ksql-utilization-metrics");
+      final MetricName nodeTotal = metricRegistry.metricName("node-storage-total", "ksql-utilization-metrics");
+      final MetricName nodeUsed = metricRegistry.metricName("node-storage-used", "ksql-utilization-metrics");
+      registeredNodeMetrics.add(nodeAvailable);
+      registeredNodeMetrics.add(nodeTotal);
+      registeredNodeMetrics.add(nodeUsed);
+
+      metricRegistry.addMetric(
+        nodeAvailable,
+        (Gauge<Long>) (config, now) -> baseDir.getFreeSpace()
+      );
+      metricRegistry.addMetric(
+        nodeTotal,
+        (Gauge<Long>) (config, now) -> baseDir.getFreeSpace()
+      );
+      metricRegistry.addMetric(
+        nodeUsed,
+        (Gauge<Long>) (config, now) -> baseDir.getFreeSpace()
+      );
+    }
   }
 
   @Override
-  public void run() {
-    logger.info("Reporting Observability Metrics");
-    final Long currentTime = time.milliseconds();
-    // here is where we would report metrics
-    lastSampleTime = currentTime;
+  public void metricChange(final KafkaMetric metric) {
+    if (!metric.metricName().name().equals("total-sst-files-size")) {
+      return;
+    }
+    final String taskId = metric.metricName().tags().getOrDefault("task-id", "");
+    final String queryIdTag = metric.metricName().tags().getOrDefault("thread-id", "");
+    Pattern pattern = Pattern.compile("(?<=query_)(.*?)(?=-)");
+    Matcher matcher = pattern.matcher(queryIdTag);
+    final String queryId = matcher.find() ? matcher.group(1) : "";
+    // if we haven't seen a task for this query yet
+    if (!metricsSeen.containsKey(queryId)) {
```

Comment:
everything from this line down to the end of this method needs to be in a synchronized block. I wouldn't synchronize this whole method though because most of the time it's called it doesn't need to do anything (since most metrics will fail the initial check).

Reply:
Synchronized on `metricsSeen`, not `metricsRegistry`, right?